### DISCUSSION TITLE
issue: Remove feature not working properly on useFieldArray

### DISCUSSION THREAD

**Discussion by charmingjae** [react-hook-form/react-hook-form#11609] [Q&A]
### Version Number

codesandbox: 7.51.0 , local: 7.50.1

### Codesandbox/Expo snack

[Codesandbox Link](https://codesandbox.io/p/devbox/4jpdq6?file=%2Fsrc%2Fcomponent%2FEditRow%2Findex.jsx%3A25%2C81&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522cltiau9k40006356ljrkiujis%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522cltiau9k40002356l9cxnj85k%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522cltiau9k40004356l8fe9gfvp%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522cltiau9k40005356lfkjbjmpi%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cltiau9k40002356l9cxnj85k%2522%253A%257B%2522id%2522%253A%2522cltiau9k40002356l9cxnj85k%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cltibn59t0002356lc5d8c3u1%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A17%252C%2522startColumn%2522%253A23%252C%2522endLineNumber%2522%253A17%252C%2522endColumn%2522%253A23%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252Fcomponent%252FInput%252Findex.jsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522cltibosfi0002356lpvzfb3aa%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A23%252C%2522startColumn%2522%253A19%252C%2522endLineNumber%2522%253A23%252C%2522endColumn%2522%253A19%257D%255D%252C%2522filepath%2522%253A%2522%252Fpackage.json%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522cltibw2io0002356lleukyiah%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A13%252C%2522startColumn%2522%253A11%252C%2522endLineNumber%2522%253A13%252C%2522endColumn%2522%253A11%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252Fpage%252FTestPage%252Findex.jsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522cltibylyq0002356mgqvs06jh%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A22%252C%2522startColumn%2522%253A16%252C%2522endLineNumber%2522%253A22%252C%2522endColumn%2522%253A16%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252Fcomponent%252FAddRow%252Findex.jsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522cltic7kwe0002356layxq7947%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A12%252C%2522startColumn%2522%253A19%252C%2522endLineNumber%2522%253A12%252C%2522endColumn%2522%253A19%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.jsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clticdvcr0002356liu33yndy%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A16%252C%2522startColumn%2522%253A6%252C%2522endLineNumber%2522%253A16%252C%2522endColumn%2522%253A6%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252FApp.jsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clticsx300002356lr360zppm%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A25%252C%2522startColumn%2522%253A81%252C%2522endLineNumber%2522%253A25%252C%2522endColumn%2522%253A81%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252Fcomponent%252FEditRow%252Findex.jsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clticsx300002356lr360zppm%2522%257D%252C%2522cltiau9k40005356lfkjbjmpi%2522%253A%257B%2522id%2522%253A%2522cltiau9k40005356lfkjbjmpi%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A5173%252C%2522id%2522%253A%2522cltibp7rb003h356leka6mdio%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522activeTabId%2522%253A%2522cltibp7rb003h356leka6mdio%2522%257D%252C%2522cltiau9k40004356l8fe9gfvp%2522%253A%257B%2522id%2522%253A%2522cltiau9k40004356l8fe9gfvp%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cltiau9k40003356lw2xuwbc7%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522cltiauaha000sdhf89xxr7z89%2522%257D%255D%252C%2522activeTabId%2522%253A%2522cltiau9k40003356lw2xuwbc7%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

### Steps to reproduce

Hi, this issue is about the feature to __add(append)__ and __delete(remove)__ data using __useFieldArray__.
Thank you so much for taking the time to look at this issue. I working very hard to resolve this issue and would appreciate any feedback you can provide. 

- Currently using a component in the ```Material UI (MUI)``` to receive data input.
   - Using multiple components such as Select, Autocomplete, as well as TextInput. The provided codesandbox only uses TextInput. 
- Using the ```Controller``` component to control the component without using ```register```.
- The ```control``` used in ```useFieldArray``` uses the control in ```useFormContext```. 

If look at the codesandbox provided above, we'll see that it has two main parts: 
1. __Input__, where we can enter data
2. __Button__, where we can add or delete data. 

Below is the order in which we can view issues. 

1. When you connect to codesandbox, you'll see the following elements.
![image](https://github.com/react-hook-form/react-hook-form/assets/54883521/e51ac805-9194-4eaf-96aa-08ca44c9aee0)

2. Enter a value from __1 to 3__ in the __topmost Input element__ and press the __Add__ button on the right to add a row.
A new row is created below in the order you entered the values. 
![image](https://github.com/react-hook-form/react-hook-form/assets/54883521/25852401-5bed-4eeb-b05a-0e6e3f1de983)

3. __Delete__ the top row of the new rows (the row with the number 1) to the row with the number 3. 
![image](https://github.com/react-hook-form/react-hook-form/assets/54883521/42d15c14-8e35-4d87-b3de-edebdf0147df)

4. When you try to delete the row with the number 2 entered, you can see that the row is not deleted and is changed to a value of the number '3'. 
![image](https://github.com/react-hook-form/react-hook-form/assets/54883521/818784a8-9db6-42b0-b7b0-a00750570ba0)

### Expected behaviour

__Expected behaviour__: Expecting things to function normally when adding and deleting data.

__1. With all data added.__

```text
   3           <-- Add row (Always displayed)
   
   1           <-- Edit row (Can be deleted)
   2           <-- Edit row (Can be deleted)
   3           <-- Edit row (Can be deleted)
```

__2. Delete 1 (index: 0)__

```text
   3           <-- Add row (Always displayed)
   
   2           <-- Edit row (Can be deleted)
   3           <-- Edit row (Can be deleted)
```

__3. Delete 2 (index: 0)__

```text
   3           <-- Add row (Always displayed)
   
   3           <-- Edit row (Can be deleted)
```

__4. Delete 3 (index: 0)__

```text
   3           <-- Add row (Always displayed)
```

But now, the functionality works as follows.

__1. With all data added.__

```text
   3           <-- Add row (Always displayed)
   
   1           <-- Edit row (Can be deleted)
   2           <-- Edit row (Can be deleted)
   3           <-- Edit row (Can be deleted)
```

__2. Delete 1 (index: 0)__ Okay, work properly.

```text
   3           <-- Add row (Always displayed)
   
   2           <-- Edit row (Can be deleted)
   3           <-- Edit row (Can be deleted)
```

__3. Delete 2 (index: 0)__ Not work properly.

```text
   3           <-- Add row (Always displayed)

   3           <-- Edit row (Can be deleted) - Change to 3 from 2
   3           <-- Edit row (Can be deleted)
```

### What browsers are you seeing the problem on?

Chrome

### Relevant log output

1. Fully data added

```text
Edit row data :  {test: '1', id: '93075968-6946-44ab-8f89-f8d77950f6b8'}
Edit row data :  {test: '2', id: 'e1b8e305-e7d9-4c8f-98d7-fcc0ec43ef22'}
Edit row data :  {test: '3', id: 'b288835c-e277-47eb-85e9-3c9ed5eb4245'}
```

2. Delete '1'
```text
Edit row data :  {test: '2', id: 'e1b8e305-e7d9-4c8f-98d7-fcc0ec43ef22'}
Edit row data :  {test: '3', id: 'b288835c-e277-47eb-85e9-3c9ed5eb4245'}
```

3. Delete '2' - New unintended rows added below
```text
Edit row data :  {test: '3', id: 'b288835c-e277-47eb-85e9-3c9ed5eb4245'}
Edit row data :  {test: '3', id: '76591c37-e6ca-4cb3-b79e-b82e1a46be65'}
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by devLucasMoraes** [comment]
I have the same problem as you, I racked my brain all night thinking it was some implementation error on my part

**Reply by ramiKang** [reply]
Hello! I was thinking about the same error and somehow found a solution. Please refer to the method below!

If you call a component using fields.map, put the key field.id in the component in the map.
And set the key to index in the top tag inside the component.

If you are using fields.map to directly invoke a component instead of invoking it, set the key value to field.id !

I'll explain using the sandbox above

#### If you call a component in fields.map
```
# src/page/TestPage/index.jsx
{fields.map((field, index) => {
        return (
          <EditRow key={field.id} data={field} rowIdx={index} remove={remove} />
        );
      })}
```

```
# src/component/EditRow/index.jsx
return (
    <div
      key={rowIdx} //  << Please add key
      style={{
        display: "flex",
        justifyContent: "center",
        alignItems: "center",
        flexDirection: "row",
      }}
    >
           # ... code
    </div>
```
#### If you are creating content directly within fields.fields.map 
```
# src/page/TestPage/index.jsx
{fields.map((field, index) => {
        return (
          <div key={field.id}>
               # code
          </div>
        );
      })}

```

I did this. I hope this solution will help you

**Reply by charmingjae** [reply]
@devLucasMoraes @ramiKang 

First, thank you for sharing your interest in my problem.
The component's key must be assigned a ```unique``` value. However, ```index``` is not a unique value when render in loop(map), but it works fine when I assign index to key as shown in the answer above. That's weird. Is there another solution?

**Reply by keithics** [reply]
@ramiKang you saved me days with this "solution".
